packages/atlas-service/src/main.ts

```diff
@@ -82,82 +138,143 @@ export class AtlasService {
-        res.setHeader('Location', redirectUrl.toString());
-        res.end();
-        return;
-      }
+  private static oidcPluginSyncedFromLoggerState: AtlasServiceAuthState =
+    'initial';
 
-      redirectRequestHandler(data);
-    },
-    async openBrowser({ url }) {
-      await shell.openExternal(url);
-    },
-    logger: this.oidcPluginLogger,
-  });
+  private static plugin: MongoDBOIDCPlugin;
 
   private static token: Token | null = null;
 
   private static signInPromise: Promise<Token> | null = null;
 
   private static fetch: typeof fetch = fetch;
 
+  private static secretStore: SecretStore = defaultSecretStore;
+
   private static refreshing = false;
 
   private static get clientId() {
-    if (!process.env.COMPASS_CLIENT_ID) {
+    const clientId =
+      process.env.COMPASS_CLIENT_ID_OVERRIDE || process.env.COMPASS_CLIENT_ID;
+
+    if (!clientId) {
       throw new Error('COMPASS_CLIENT_ID is required');
     }
     return process.env.COMPASS_CLIENT_ID;
   }
 
   private static get issuer() {
-    if (!process.env.COMPASS_OIDC_ISSUER) {
+    const issuer =
+      process.env.COMPASS_OIDC_ISSUER_OVERRIDE ||
+      process.env.COMPASS_OIDC_ISSUER;
+
+    if (!issuer) {
       throw new Error('COMPASS_OIDC_ISSUER is required');
     }
     return process.env.COMPASS_OIDC_ISSUER;
   }
 
   private static get apiBaseUrl() {
-    if (!process.env.COMPASS_ATLAS_SERVICE_BASE_URL) {
+    const apiBaseUrl =
+      process.env.COMPASS_ATLAS_SERVICE_BASE_URL_OVERRIDE ||
+      process.env.COMPASS_ATLAS_SERVICE_BASE_URL;
+
+    if (!apiBaseUrl) {
       throw new Error(
         'No AI Query endpoint to fetch. Please set the environment variable `COMPASS_ATLAS_SERVICE_BASE_URL`'
       );
     }
     return process.env.COMPASS_ATLAS_SERVICE_BASE_URL;
   }
 
-  static init() {
-    if (this.calledOnce) {
-      return;
+  // We use `allowedFlows` plugin option to control whether or not plugin is
+  // allowed to start sign in flow or just try refresh and fail if refresh is
+  // not possible.
+  //  - If we are authenticated, token expired, or we are restoring state, we
+  //    only allow token refresh
+  //  - Any other state: initial or error, will allow sign in using auth-code
+  //    flow in addition to token refresh
+  private static getAllowedAuthFlows(): AuthFlowType[] {
+    return ['restoring', 'authenticated', 'expired'].includes(
+      this.oidcPluginSyncedFromLoggerState
+    )
+      ? []
+      : ['auth-code'];
+  }
+
+  static init(): Promise<void> {
+    if (this.initPromise) {
+      return this.initPromise;
     }
-    this.calledOnce = true;
-    ipcExpose('AtlasService', this, [
-      'getUserInfo',
-      'introspect',
-      'isAuthenticated',
-      'signIn',
-      'getQueryFromUserInput',
-    ]);
-    this.attachOidcPluginLoggerEvents();
-    log.info(
-      mongoLogId(1_001_000_210),
-      'AtlasService',
-      'Atlas service initialized'
+    this.initPromise = (async () => {
```

Comment:
Can you do
```suggestion
    return this.initPromise ??= (async () => {
```

or would the linter complain about that?

Reply:
Oh, yes, that's way better! I don't think it will complain, I just always forget about null-ish operators!